File: kimbyulyi/PGS/Level2/삼각달팽이.java
//프로그래머스 삼각 달팽이
class Solution {
    public static int[] solution(int n) {
        int[][] arr = new int[n][n];

        int x = -1; //처음엔 무조건 아래쪽을 향하므로 -1로 초기화
        int y = 0;
        int num = 1;

        for (int i = 0; i < n; i++) {
            for (int j = i; j < n; j++) {
                if (i % 3 == 0) { // 아래쪽을 향하는 경우
                    x++;
                } else if (i % 3 == 1) { // 오른쪽을 향하는 경우
                    y++;
                } else if (i % 3 == 2){ // 대각선 위쪽을 향하는 경우
                    x--;
                    y--;
                }
                arr[x][y] = num++;
            }
        }

        int[] answer = new int[n * (n + 1) / 2]; // 삼각형의 크기
        int idx = 0;
        for (int[] a : arr) {
            for (int b : a) {
                if (b == 0) break;  //값이 0인 경우 제외해주기
                answer[idx++] = b;
            }
        }
        return answer;
    }

    public static void main(String[] args) {
        int[] arr = solution(4);
        for (int a : arr) {
            System.out.println(a);
        }
    }

File: baekyeonsu/PGS/Lv. 2/삼각 달팽이.js
function solution(n) {
  let stack = [];
  let totalNumber = 0;
  for (let i = 1; i <= n; i++) {
    stack.push(new Array(i).fill(0));
    totalNumber += i;
  }

  let [x, y] = [0, 0];
  let [dicX, dicY] = [1, 0];
  for (let num = 1; num <= totalNumber; num++) {
    stack[x][y] = num;

    if (
      dicX === 1 &&
      dicY === 0 &&
      (x + dicX === n || stack[x + dicX][y] !== 0)
    ) {
      [dicX, dicY] = [0, 1];
    }
    if (
      dicY === 1 &&
      dicX === 0 &&
      (y + dicY === n || stack[x][y + dicY] !== 0)
    ) {
      [dicX, dicY] = [-1, -1];
    }
    if (
      dicX === -1 &&
      dicY === -1 &&
      (x + dicX === -1 || stack[x + dicX][y + dicY] !== 0)
    ) {
      [dicX, dicY] = [1, 0];
    }
    x += dicX;
    y += dicY;
  }

  return stack.reduce((acc, cur) => acc.concat(cur), []);
}
